serviceaccount/reconciler.go

```diff
@@ -0,0 +1,148 @@
+package serviceaccount
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/giantswarm/to"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/giantswarm/workload-identity-operator-gcp/webhook"
+)
+
+const (
+	AnnotationSecretMetadata = "kubernetes.io/service-account.name" //#nosec G101
+
+	SecretNameSuffix = "google-application-credentials" //#nosec G101
+)
+
+type ServiceAccountReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	Logger logr.Logger
+}
+
+//+kubebuilder:rbac:groups=core,resources=serviceaccounts,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core,resources=serviceaccounts/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=core,resources=serviceaccounts/finalizers,verbs=update
+
+func (r *ServiceAccountReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	serviceAccount := &corev1.ServiceAccount{}
+
+	err := r.Client.Get(ctx, req.NamespacedName, serviceAccount)
+	if err != nil {
+		r.Logger.Error(err, "could not get service account", "service account", req.NamespacedName)
+		return reconcile.Result{}, nil
+	}
+
+	gcpServiceAccount, isGCPAnnotated := serviceAccount.Annotations[webhook.AnnotationGCPServiceAccount]
+	workloadIdentityPool, hasWorkloadIdentity := serviceAccount.Annotations[webhook.AnnotationWorkloadIdentityPoolID]
+	identityProvider, hasIdentityProvider := serviceAccount.Annotations[webhook.AnnotationGCPIdentityProvider]
+
+	if !isGCPAnnotated {
+		message := fmt.Sprintf("ServiceAccount misssing %q annotation", webhook.AnnotationGCPServiceAccount)
+		err = errors.New(message)
+		r.Logger.Error(err, message, "service-account", req.NamespacedName)
```

Comment:
This controller is watching all `ServiceAccounts` right? It feels weird to log an error for every `ServiceAccount` that doesn't contain the annotation expected by this controller. Maybe logging that we are skipping a `ServiceAccount` with a `info` message would make more sense? What do you think?

Reply:
That makes more sense and will cause less noise. Thanks 👍🏾